src/pip/_internal/req/req_uninstall.py

```diff
@@ -422,154 +407,166 @@ def rollback(self) -> None:
 
-        elif egg_info_exists and dist.egg_info.endswith(".dist-info"):
+        elif dist.installed_with_dist_info:
             for path in uninstallation_paths(dist):
                 paths_to_remove.add(path)
 
         elif develop_egg_link:
-            # develop egg
+            # PEP 660 modern editable is handled in the ``.dist-info`` case
+            # above, so this only covers the setuptools-style editable.
             with open(develop_egg_link) as fh:
                 link_pointer = os.path.normcase(fh.readline().strip())
-            assert (
-                link_pointer == dist.location
-            ), "Egg-link {} does not match installed location of {} (at {})".format(
-                link_pointer, dist.project_name, dist.location
+            assert link_pointer == dist_location, (
+                f"Egg-link {link_pointer} does not match installed location of "
+                f"{dist.raw_name} (at {dist_location})"
             )
             paths_to_remove.add(develop_egg_link)
             easy_install_pth = os.path.join(
                 os.path.dirname(develop_egg_link), "easy-install.pth"
             )
-            paths_to_remove.add_pth(easy_install_pth, dist.location)
+            paths_to_remove.add_pth(easy_install_pth, dist_location)
 
         else:
             logger.debug(
                 "Not sure how to uninstall: %s - Check: %s",
                 dist,
-                dist.location,
+                dist_location,
             )
 
+        if dist.in_usersite:
+            bin_dir = get_bin_user()
+        else:
+            bin_dir = get_bin_prefix()
+
         # find distutils scripts= scripts
-        if dist.has_metadata("scripts") and dist.metadata_isdir("scripts"):
-            for script in dist.metadata_listdir("scripts"):
-                if dist_in_usersite(dist):
-                    bin_dir = get_bin_user()
-                else:
-                    bin_dir = get_bin_prefix()
-                paths_to_remove.add(os.path.join(bin_dir, script))
+        try:
+            for script in dist.iterdir("scripts"):
+                paths_to_remove.add(os.path.join(bin_dir, script.name))
                 if WINDOWS:
-                    paths_to_remove.add(os.path.join(bin_dir, script) + ".bat")
-
-        # find console_scripts
-        _scripts_to_remove = []
-        console_scripts = dist.get_entry_map(group="console_scripts")
-        for name in console_scripts.keys():
-            _scripts_to_remove.extend(_script_names(dist, name, False))
-        # find gui_scripts
-        gui_scripts = dist.get_entry_map(group="gui_scripts")
-        for name in gui_scripts.keys():
-            _scripts_to_remove.extend(_script_names(dist, name, True))
-
-        for s in _scripts_to_remove:
+                    paths_to_remove.add(os.path.join(bin_dir, f"{script.name}.bat"))
+        except (FileNotFoundError, NotADirectoryError):
```

Comment:
Let's reduce the scope of this try-except, to the bit that can actually raise these errors, which I imagine is the `dist.iterdir` call?

Reply:
Hmm, this is a bit complicated because `iterdir()` does not raise when invoked, only when the return value is iterated on:

```pycon
>>> import pathlib
>>> p = pathlib.Path('asd')
>>> p.exists()
False
>>> it = p.iterdir()
>>> next(it)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\uranusjr\AppData\Local\Programs\Python\Python310\lib\pathlib.py", line 1015, in iterdir
    for name in self._accessor.listdir(self):
FileNotFoundError: [WinError 3] The system cannot find the path specified: 'asd'
```

So there’s really not a good way to make the scope smaller without some elaborated setup, e.g.

```python
def _safe_iterate(it):
    while True:
        try:
            yield next(it)
        except (FileNotFoundError, NotADirectoryError):
            return
        except Stopiteration:
            return
```